Update sdba documentation (#1230)
Ouranosinc/xclim: docs/sdba.rst

Bias adjustment and downscaling algorithms
==========================================

> The `xclim.sdba` submodule provides bias-adjustment methods and will eventually provide statistical downscaling algorithms.

Reviewer: ```suggestion
The `xclim.sdba` submodule provides a collection of bias-adjustment methods meant to meant to correct for systematic biases found in climate model simulations relative to observations. 
```

> Almost all adjustment algorithms conform to the `train` - `adjust` scheme, formalized within `TrainAdjust` classes.

Reviewer: ```suggestion
Almost all adjustment algorithms conform to the `train` - `adjust` scheme, meaning that adjustment factors are first estimated on training data sets, then applied in a distinct step to the data to be adjusted.  
```

Given a reference time series (ref), historical simulations (hist) and simulations to be adjusted (sim),
any bias-adjustment method would be applied by first estimating the adjustment factors between the historical simulation
and the observation series, and then applying these factors to `sim`, which could be a future simulation::

  # Create the adjustment object by training it with reference and model data, plus certain arguments
  Adj = Adjustment.train(ref, hist, group="time.month")
  # Get a scenario by applying the adjustment to a simulated timeseries.
  scen = Adj.adjust(sim, interp="linear")
  Adj.ds.af  # adjustment factors.

> Most method can either be applied additively by multiplication.

Reviewer: ```suggestion
Most method support both additive and multiplicative correction factors.
```

Also, the `group` argument allows adjustment factors to be estimated independently for different periods: the full

> time series,  months, seasons or day of the year. The `interp` argument then allows for interpolation between these

Reviewer: ```suggestion
time series,  months, seasons or day of the year.  For monthly groupings, the `interp` argument then allows for interpolation between
```

> adjustment factors to avoid discontinuities in the bias-adjusted series (only applicable for monthly grouping).

Reviewer: ```suggestion
adjustment factors to avoid discontinuities in the bias-adjusted series.
```

series, and possibly reduce the number of quantile bins that needs to be used.

Modular Approach
================

The module atempts to adopt a modular approach instead of implementing published and named methods directly.
A generic bias adjustment process is laid out as follows:

- preprocessing on ``ref``, ``hist`` and ``sim`` (using methods in :py:mod:`xclim.sdba.processing` or :py:mod:`xclim.sdba.detrending`)
- creating and training the adjustment object ``Adj = Adjustment.train(obs, sim, **kwargs)`` (from :py:mod:`xclim.sdba.adjustment`)
- adjustment ``scen = Adj.adjust(sim, **kwargs)``
- post-processing on ``scen`` (for example: re-trending)

The train-adjust approach allows to inspect the trained adjustment object. The training information is stored in
the underlying `Adj.ds` dataset and usually has a `af` variable with the adjustment factors. Its layout and the
other available variables vary between the different algorithm, refer to :ref:`Adjustment methods <sdbauserapi>`.

Parameters needed by the training and the adjustment are saved to the ``Adj.ds`` dataset as a `adj_params` attribute.

> Other parameters, those only needed by the adjustment are passed in the `adjust` call and written to the history

Reviewer: unclear

attribute in the output scenario DataArray.

.. _grouping:

Grouping
========

For basic time period grouping (months, day of year, season), passing a string to the methods needing it is sufficient.
Most methods acting on grouped data also accept a `window` int argument to pad the groups with data from adjacent ones.
Units of `window` are the sampling frequency of the main grouping dimension (usually `time`). For more complex grouping,

> one can pass an instance of :py:class:`xclim.sdba.base.Grouper` directly.

Reviewer: I think it would help to see and example here of a grouping declaration.

In theory, short wave radiation should be capped when precipitation is not zero, but there is as of yet no mechanism proposed to do that, see :cite:t:`hoffmann_meteorologically_2012`.

> Other documentation

Reviewer: ```suggestion
Usage examples
```

===================
The usage of this module is documented in two example notebooks: `SDBA <notebooks/sdba.ipynb>`_ and `SDBA advanced <notebooks/sdba-advanced.ipynb>`_.

> Some issues were also discussed on the Github repository. Most of these are still open questions, feel free to participate to the discussion!

Reviewer: Discussion topics
==============

In theory, short wave radiation should be capped when precipitation is not zero, but there is as of yet no mechanism proposed to do that, see :cite:t:`hoffmann_meteorologically_2012`.

Usage examples

> ===================

Reviewer 2: ```suggestion
==============
```
